### PR TITLE
Changes to Solrize re-index stage to minimise time docs are absent

### PR DESCRIPTION
**Pull request by simon-20**

### CODE REVIEW

File: src/library/solrize.py [hunk 1]
```diff
@@ -126,69 +168,38 @@ def process_hash_list(document_datasets):
...
-                    raise SolrError('DELETING hash: ' + file_hash + ' and id: ' + file_id +
-                                    ', from collection with name ' + core_name + ': ' + e_message)
-
-            logger.info('Adding docs for hash: ' + file_hash + ' and id: ' + file_id)
+            # check whether existing DS data and new dataset contain only "good" data:
+            # - flattened activities have any duplicated IATI identifiers in them?
+            # - existing docs in solr have repeated IDs
+            new_identifiers = [fa['iati_identifier'] for fa in flattened_activities[0]]
+            has_dupes = len(set(new_identifiers)) != len(new_identifiers)
+            has_dupes = has_dupes or solr_cores['activity'].search("id:*--1", rows=0).hits > 0
```
Comment (odscjames) [inline src/library/solrize.py:176]: I don't understand how this works - doesn't it need to search limited by the file-id too?

File: src/library/solrize.py [hunk 2]
```diff
@@ -96,6 +96,48 @@ def process_hash_list(document_datasets):
         doc.solr_api_error
     """
 
+    def delete_from_solr(cores: dict, file_id: str, file_hash: str, query: dict):
```
Comment (odscjames) [inline src/library/solrize.py:99]: Is there a reason this and the following one are defined in process_hash_list and not just at the top level?

Comment (simon-20) [reply src/library/solrize.py:99]: I put them here since there are only used here at the moment.

File: src/library/solrize.py (hunk 1 above)
Comment (simon-20) [reply src/library/solrize.py:176]: Yeah, whoops. I had that check in and lost it somewhere along the lines in a refactor.

File: src/library/solrize.py (hunk 2 above)
Comment (odscjames) [reply src/library/solrize.py:99]: Hmm .... I think you could say that about a lot of code in refresher, and it's all at top level. And maybe we'll write tests one day (tho these funcs are probably lower on the list of priority candidates). This way left me wondering if there was a special reason for it. I think I'd prefer it at top level but that's not a strongly held view. 

(Also, will it be faster to have them defined outside a loop so Python only defines them once? Tho probably very little difference there compared to other speedups we could be working on)

Comment (simon-20) [reply src/library/solrize.py:99]: No special reason, just wanted to keep top level clean as possible. But makes sense to move it if we might refactor further and make tests at some point.

**Comment by odscjames** [comment]
Looks good, thanks!

**Pull request merged**